Fix bug with new nested where filters

tjhiggins opened the pull request:
### Description of change
Found a bug in the most recent nested where filter feature:
https://github.com/typeorm/typeorm/pull/7805

When combining nested primary keys and non-primary keys the query builder would build an incorrect query.
Included a test case that illustrates the issue. 

 fixes #8082

### Pull-Request Checklist

- [x] Code is up-to-date with the `master` branch
- [x] `npm run lint` passes with this change
- [x] `npm run test` passes with this change
- [x] There are new or updated unit tests validating the change
- [x] The new commits follow conventions explained in [CONTRIBUTING.md](https://github.com/typeorm/typeorm/blob/master/CONTRIBUTING.md)

tjhiggins commented:
@imnotjames how do I check if an entity is a transformed class vs a plain filter object? struggling with the differences between:

```ts
const foundChild = await manager.findOne(Child, { parent: loadedParent });
```

and 

```ts
.where({
    post: {
        category: {
            id: In([1, 2, 3]),
            name: In(["Foo", "Bar", "Baz"]),
        }
    }
})
```

tjhiggins commented:
@imnotjames The node 14 test failed due to a circleci issue - could you please re-run?

imnotjames commented:
Doesn't the test pass if you write the code as

```
const isEntityTarget = relation.inverseEntityMetadata.target === entity[key].constructor;

if (isEntityTarget) {
    // .... shortcuts for entity target
}

// If nothing else, just return every property that's being passed to us.
const subPaths = this.createPropertyPath(relation.inverseEntityMetadata, entity[key])
    .map(p => `${path}.${p}`);
paths.push(...subPaths);
```

tjhiggins commented:
> Doesn't the test pass if you write the code as
> 
> ```
> const isEntityTarget = relation.inverseEntityMetadata.target === entity[key].constructor;
> 
> if (isEntityTarget) {
>     // .... shortcuts for entity target
> }
> 
> // If nothing else, just return every property that's being passed to us.
> const subPaths = this.createPropertyPath(relation.inverseEntityMetadata, entity[key])
>     .map(p => `${path}.${p}`);
> paths.push(...subPaths);
> ```

I tried that after I added the isEntityTarget check, but a couple tests failed. Realizing now that those tests might be invalid now and it now generates the correct sql. Let me know if I'm wrong:

https://github.com/typeorm/typeorm/pull/8005/commits/6de932714bf4fa1f6fa112b5e143a892d16f069c#diff-fc53d8864169b2e2d0b175dc446c164524f59407c457d6b81a0fa1c004d8ca5cR206

Updated with the simple solution and changed those tests.

tjhiggins commented:
Nvm the simple solution doesn't play nice with updates:
```bash
1) embedded > embedded-many-to-one-case1
       owner side
         should insert, load, update and remove entities with embeddeds when embedded entity having ManyToOne relation2:
     EntityColumnNotFound: No entity column "counters.likedUser.id" was found.
```

tjhiggins commented:
@imnotjames Ready for another look. Would love your feedback on a better check for hasOnlyPrimaryKeys. Also please note that a couple tests that used to fail now pass - I'm not 100% if they are correct.

imnotjames commented on test/functional/connection-options-reader/connection-options-reader.ts:
```diff
@@ -11,7 +11,7 @@ async function createDotenvFiles() {
 
 async function createYamlFiles() {
-  await fs.mkdir(path.join(__dirname, "configs/yaml"));
+  await fs.mkdir(path.join(__dirname, "configs/yaml"), { recursive: true });
```
Why is this change needed?

tjhiggins replied:
I wasn't able to run test-fast multiple times in a row. It would complain that the directory already existed.

imnotjames replied:
That's a bigger problem that our tests have than just this.  Can you omit this change for now just to simplify this PR & what it touches?  If we have to revert or find the cause of something it makes it a whole lot easier if it's one PR for one change.

imnotjames commented on src/query-builder/QueryBuilder.ts:
```diff
@@ -993,51 +993,68 @@ export abstract class QueryBuilder<Entity> {
+                        }
+                    }
+
+                    if (relation.relationType === "one-to-many" || relation.relationType === "many-to-many") {
+                        throw new Error(`Cannot query across ${relation.relationType} for property ${path}`);
+                    }
```
Pretty sure this needs to remain outside the `if (isEntityTarget || hasOnlyPrimaryKeys) { ... }` block as it's unrelated to the shortcuts taken.  We just do not currently support this kind of query at all.

imnotjames commented on src/query-builder/QueryBuilder.ts:
```diff
@@ -993,51 +993,68 @@ export abstract class QueryBuilder<Entity> {
+                    // These handles the situation where someone passes the model & we don't need to make
+                    // a HUGE `where` to uniquely look up the entity.
+
+                    // In the case of a *-to-one, there's only ever one possible entity on the other side
+                    // so if the join columns are all defined we can return just the relation itself
+                    // because it will fetch only the join columns and do the lookup.
+                    if (relation.relationType === "one-to-one" || relation.relationType === "many-to-one") {
+                        const joinColumns = relation.joinColumns
+                            .map(j => j.referencedColumn)
+                            .filter((j): j is ColumnMetadata => !!j);
+
+                        const hasAllJoinColumns = joinColumns.length > 0 && joinColumns.every(
+                            column => column.getEntityValue(entity[key], false)
+                        );
+
+                        if (hasAllJoinColumns) {
+                            paths.push(path);
+                            continue;
+                        }
+                    }
+
```
Does the `hasOnlyPrimaryKeys` actually apply here?  Isn't it just in the case of an `isEntityTarget`?
  Can this ever apply with `hasOnlyPrimaryKeys`?

tjhiggins replied:
I believe you're right and I can have this block only apply to isEntityTarget.

imnotjames commented on src/query-builder/QueryBuilder.ts:
```diff
@@ -993,51 +993,68 @@ export abstract class QueryBuilder<Entity> {
+                        if (value) {
+                            seenPrimaryColumns.add(columnName);
+                        } else {
+                            break;
```
If we know this inner one is invalid, we know they all are.  Do we want to `break` further up?

imnotjames commented on src/query-builder/QueryBuilder.ts:
```diff
@@ -993,51 +993,68 @@ export abstract class QueryBuilder<Entity> {
+                const seenPrimaryColumns = new Set<string>();
+                for (const primaryColumn of primaryColumns) {
+                    const entityValueMap = primaryColumn.getEntityValueMap(entity[key]);
+                    for (const [columnName, value] of Object.entries(entityValueMap || {})) {
```
What about embedded entity primary keys?  Is that a concern here?

tjhiggins replied:
I'm not sure - happy to hand off this pr to you if its easier. 

Otherwise please point me to docs and I can try to add a test case.

imnotjames commented on src/query-builder/QueryBuilder.ts:
```diff
@@ -993,51 +993,68 @@ export abstract class QueryBuilder<Entity> {
+                for (const primaryColumn of primaryColumns) {
+                    const entityValueMap = primaryColumn.getEntityValueMap(entity[key]);
+                    for (const [columnName, value] of Object.entries(entityValueMap || {})) {
+                        if (value) {
```
Is `null` a valid PK value?  I know `0` is.  So is an empty string.  Is that something we should be handling here?  Only checking for `undefined` instead?

imnotjames commented on src/query-builder/QueryBuilder.ts:
```diff
@@ -993,51 +993,68 @@ export abstract class QueryBuilder<Entity> {
+                        }
+                    }
+                }
+                const hasOnlyPrimaryKeys = primaryColumns.length > 0 && seenPrimaryColumns.size === Object.keys(entity[key]).length;
```
If it's only primary columns do we actually care if it hits the optimizations?

In theory, it would be the same either way, wouldn't it?

tjhiggins replied:
The update flow uses this method and fails if you pass `contact.id` vs `contactId`. 
ex. https://github.com/typeorm/typeorm/pull/8005#issuecomment-891214828
That's the reason I added the hasOnlyPrimaryKeys check back.

pleerock commented:
Closing in favour of #8082
